coffee/visitors/inspectors.py

```diff
@@ -0,0 +1,540 @@
+        super(CountOccurences, self).__init__()
+
+    def visit_object(self, o, env):
+        # Not a symbol, return identity for summation
+        return {}
+
+    def visit_list(self, o, env):
+        # Walk list entries (since some operands methods return lists)
+        ret = Counter()
+        for entry in o:
+            ret.update(self.visit(entry, env=env))
+        return ret
+
+    def visit_Node(self, o, env, *args, **kwargs):
+        # Walk the transformed children and sum up.
+        ret = Counter()
+        for d in args:
+            ret.update(d)
+        return ret
+
+    def visit_Assign(self, o, env):
+        ret = Counter()
+        if self.rvalues:
+            # Only counting rvalues, so don't walk lvalue
+            ops = o.children[1:]
+        else:
+            ops = o.children
+        args = [self.visit(op, env=env) for op in ops]
+        for d in args:
+            ret.update(d)
+        return ret
+
+    visit_Incr = visit_Assign
+
+    visit_Decr = visit_Assign
+
+    def visit_Symbol(self, o, env):
+        return {self.key(o): 1}
+
+
+class DetermineUnrollFactors(Visitor):
+    """
+    Determine unroll factors for all For loops in a tree.
+
+    Returns a dict mapping iteration variable names to possible unroll
+    factors for that iteration variable.
+
+    Innermost loops always get an unroll factor of 1, to give the
+    backend compiler a chance of auto-vectorizing them.  Outer loops
+    are given potential unroll factors that result in no loop
+    remainders.
+    """
+    def visit_object(self, o, env):
+        return {}
+
+    def visit_Node(self, o, env, *args, **kwargs):
+        ret = {}
+        for a in args:
+            ret.update(a)
+        return ret
+
+    def visit_For(self, o, env):
+        ret = self.visit(o.children[0], env=env)
+        if len(ret) is 0:
+            # No child for loops
+            # Inner loops are not unrollable
+            ret[o.dim] = (1, )
+            return ret
+        # Not an inner loop, determine unroll factors
+        ret[o.dim] = tuple(i for i in range(1, o.size+1) if (o.size % i) == 0)
+        return ret
+
+
+class MaxLoopDepth(Visitor):
+
+    """Return the maximum loop depth in the tree."""
+
+    def visit_object(self, o, env):
+        return 0
+
+    def visit_Node(self, o, env, *args, **kwargs):
+        if len(args) == 0:
+            return 0
+        return max(args)
+
+    def visit_For(self, o, env, *args, **kwargs):
+        return 1 + max(args)
+
+
+class FindLoopNests(Visitor):
+
+    """Return a list of lists of loop nests in the tree.
+
+    Each list entry describes a loop nest with the outer-most loop
+    first.  Each entry therein is a tuple (loop_node, node_parent).
+
+    By default the top-level call to visit will record a node_parent
+    of None for the visited Node.  To provide one, pass an
+    environment in to the visitor::
+
+    .. code-block::
+
+       v.visit(node, {node_parent: parent})
```

Comment:
`v.visit(node, {'node_parent': parent})`? (Quotes around `node_parent`.)
You probably want replace all, this is not the only such documentation comment.